image-hd.c

```diff
@@ -526,10 +526,21 @@ static int hdimage_insert_gpt(struct image *image, struct list_head *partitions)
 
 		i++;
 	}
-	if (smallest_offset == ~0ULL)
-		smallest_offset = hd->gpt_location + (GPT_SECTORS - 1)*512;
-	header.first_usable_lba = htole64(smallest_offset / 512);
+	/*
+	 * Find the last non-partition data before the first partition.
+	 * This can be a bootloader of the GPT partition table.
+	 * */
+	list_for_each_entry(part, partitions, list) {
+		unsigned long long end;
 
+		if (part->in_partition_table)
+			continue;
+
+		end = part->offset + part->size;
+		if (end <= smallest_offset && end > extra_offset)
+			extra_offset = end;
+	}
+	header.first_usable_lba = htole64(roundup(extra_offset, 512) / 512);
```

Comment:
Overall, seems rather sane, and certainly much simpler than the thing I had started. There's one corner case: If there are no partitions ->in_partition_table, we'll (still) have smallest_offset = ~0ULL, so all non-partition data, including the backup table, will be taken into account, and we'll end up setting first_usable_lba to the end of the disk AFAICT. I'm not sure how to sanely handle that case; in the code you remove, it was simply set to end of gpt array.

Reply:
Is there some way this computation could easily be moved to _setup instead of being done during _generate? I'd still like to keep the door open for allowing first_usable_offset to be set explicitly, with some sanity checking during setup, but of course such a refactoring could also be done later.

Reply:
I've added a check to skip ignore the backup table. It's probably not a big use-case, what's the use of an image without content, but better get it right for correctness.

And if we ever add an option to explicitly set the value then we can skip the calculation here. And zero is never valid anyways, so we can use that for the 'not set' default.